File: backend/tests/factories/target_factory.py
"""Target Factory - Generate random test data using Faker.

This factory creates random but valid Target objects for testing.
All generated values respect the validation constraints.
"""

import uuid
from faker import Faker

from src.models.target import Target, TargetCreate, TargetUpdate
from src.generated.dtos import TargetDTO, TargetCreateDTO
from src.dal.entities.target_entity import TargetEntity

fake = Faker()

# Common frequency values for realistic data
COMMON_FREQUENCIES = [433.0, 915.0, 2.4, 5.2, 5.8]


def _random_latitude() -> float:
    """Generate random latitude between -90 and 90."""
    return round(fake.pyfloat(min_value=-90, max_value=90), 6)


def _random_longitude() -> float:
    """Generate random longitude between -180 and 180."""
    return round(fake.pyfloat(min_value=-180, max_value=180), 6)


def _random_altitude() -> float:
    """Generate random altitude (can be negative for below sea level)."""
    return round(fake.pyfloat(min_value=-500, max_value=50000), 2)


def _random_frequency() -> float:
    """Generate random frequency (positive number)."""
    # 70% chance of common frequency, 30% chance of random
    if fake.random.random() < 0.7:
        return fake.random.choice(COMMON_FREQUENCIES)
    return round(fake.pyfloat(min_value=0.1, max_value=100), 2)


def _random_speed() -> float:
    """Generate random speed (non-negative)."""
    return round(fake.pyfloat(min_value=0, max_value=1000), 2)


def _random_bearing() -> float:
    """Generate random bearing between 0 and 360."""
    return round(fake.pyfloat(min_value=0, max_value=360), 2)


def _random_ip_address() -> str:
    """Generate random valid IPv4 address."""
    return fake.ipv4()


def create_random_target(id: str | None = None) -> Target:
    """Create a random Target plain object.
    
    Args:
        id: Optional UUID string. If None, generates a new one.
        
    Returns:
        Random Target object with valid values
    """
    return Target(
        id=id or str(uuid.uuid4()),
        latitude=_random_latitude(),
        longitude=_random_longitude(),
        altitude=_random_altitude(),
        frequency=_random_frequency(),
        speed=_random_speed(),
        bearing=_random_bearing(),
        ip_address=_random_ip_address(),
    )


def create_random_target_create() -> TargetCreate:
    """Create a random TargetCreate plain object.
    
    Returns:
        Random TargetCreate object with valid values
    """
    return TargetCreate(
        latitude=_random_latitude(),
        longitude=_random_longitude(),
        altitude=_random_altitude(),
        frequency=_random_frequency(),
        speed=_random_speed(),
        bearing=_random_bearing(),
        ip_address=_random_ip_address(),
    )


def create_random_target_update(partial: bool = True) -> TargetUpdate:
    """Create a random TargetUpdate plain object.
    
    Args:
        partial: If True, only some fields are set. If False, all fields are set.
        
    Returns:
        Random TargetUpdate object
    """
    if partial:
        # Randomly select which fields to update
        update = TargetUpdate()
        if fake.random.random() < 0.5:
            update.latitude = _random_latitude()
        if fake.random.random() < 0.5:
            update.longitude = _random_longitude()
        if fake.random.random() < 0.5:
            update.altitude = _random_altitude()
        if fake.random.random() < 0.5:
            update.frequency = _random_frequency()
        if fake.random.random() < 0.5:
            update.speed = _random_speed()
        if fake.random.random() < 0.5:
            update.bearing = _random_bearing()
        if fake.random.random() < 0.5:
            update.ip_address = _random_ip_address()
        
        # Ensure at least one field is set
        if not update.has_updates():
            update.speed = _random_speed()
        
        return update
    else:
        return TargetUpdate(
            latitude=_random_latitude(),
            longitude=_random_longitude(),
            altitude=_random_altitude(),
            frequency=_random_frequency(),
            speed=_random_speed(),
            bearing=_random_bearing(),
            ip_address=_random_ip_address(),
        )


def create_random_target_dto(id: str | None = None) -> TargetDTO:
    """Create a random TargetDTO.
    
    Args:
        id: Optional UUID string. If None, generates a new one.
        
    Returns:
        Random TargetDTO with valid values
    """
    return TargetDTO(
        id=id or str(uuid.uuid4()),
        latitude=_random_latitude(),
        longitude=_random_longitude(),
        altitude=_random_altitude(),
        frequency=_random_frequency(),
        speed=_random_speed(),
        bearing=_random_bearing(),
        ip_address=_random_ip_address(),
    )


def create_random_target_create_dto() -> TargetCreateDTO:
    """Create a random TargetCreateDTO.
    
    Returns:
        Random TargetCreateDTO with valid values
    """
    return TargetCreateDTO(
        latitude=_random_latitude(),
        longitude=_random_longitude(),
        altitude=_random_altitude(),
        frequency=_random_frequency(),
        speed=_random_speed(),
        bearing=_random_bearing(),
        ip_address=_random_ip_address(),
    )


def create_random_target_entity(id: str | None = None) -> TargetEntity:
    """Create a random TargetEntity.
    
    Args:
        id: Optional UUID string. If None, generates a new one.
        
    Returns:
        Random TargetEntity with valid values (all fields as strings)
    """
    return TargetEntity(
        id=id or str(uuid.uuid4()),
        latitude=str(_random_latitude()),
        longitude=str(_random_longitude()),
        altitude=str(_random_altitude()),
        frequency=str(_random_frequency()),
        speed=str(_random_speed()),
        bearing=str(_random_bearing()),
        ip_address=_random_ip_address(),
    )


def create_random_targets(count: int = 5) -> list[Target]:
    """Create a list of random Target objects.
    
    Args:
        count: Number of targets to create
        
    Returns:
        List of random Target objects
    """
    return [create_random_target() for _ in range(count)]

File: frontend/src/test/factories/targetFactory.ts
/**
 * Target Factory - Generate random test data using Faker.
 *
 * This factory creates random but valid Target objects for testing.
 * All generated values respect the validation constraints.
 * Types are imported from generated OpenAPI code.
 *
 * Used by:
 * - Unit tests (Vitest)
 * - E2E tests (Cypress)
 * - MSW mock handlers
 */

import { faker } from "@faker-js/faker";
import type { TargetDTO, TargetCreateDTO, TargetUpdateDTO } from "../../generated/api";

// Re-export types for convenience
export type { TargetDTO, TargetCreateDTO, TargetUpdateDTO };

// Common frequency values for realistic data
const COMMON_FREQUENCIES = [433, 915, 2.4, 5.2, 5.8];

/**
 * Generate a random latitude between -90 and 90
 */
function randomLatitude(): number {
  return parseFloat(faker.location.latitude().toFixed(6));
}

/**
 * Generate a random longitude between -180 and 180
 */
function randomLongitude(): number {
  return parseFloat(faker.location.longitude().toFixed(6));
}

/**
 * Generate a random altitude
 */
function randomAltitude(): number {
  return parseFloat(faker.number.float({ min: -500, max: 50000 }).toFixed(2));
}

/**
 * Generate a random frequency (positive number)
 */
function randomFrequency(): number {
  // 70% chance of common frequency, 30% chance of random
  if (Math.random() < 0.7) {
    return faker.helpers.arrayElement(COMMON_FREQUENCIES);
  }
  return parseFloat(faker.number.float({ min: 0.1, max: 100 }).toFixed(2));
}

/**
 * Generate a random speed (non-negative)
 */
function randomSpeed(): number {
  return parseFloat(faker.number.float({ min: 0, max: 1000 }).toFixed(2));
}

/**
 * Generate a random bearing between 0 and 360
 */
function randomBearing(): number {
  return parseFloat(faker.number.float({ min: 0, max: 360 }).toFixed(2));
}

/**
 * Generate a random valid IPv4 address
 */
function randomIpAddress(): string {
  return faker.internet.ipv4();
}

/**
 * Create a random TargetDTO
 */
export function createRandomTarget(id?: string): TargetDTO {
  return {
    id: id ?? faker.string.uuid(),
    latitude: randomLatitude(),
    longitude: randomLongitude(),
    altitude: randomAltitude(),
    frequency: randomFrequency(),
    speed: randomSpeed(),
    bearing: randomBearing(),
    ip_address: randomIpAddress(),
  };
}

/**
 * Create a random TargetCreateDTO (no ID)
 */
export function createRandomTargetCreate(): TargetCreateDTO {
  return {
    latitude: randomLatitude(),
    longitude: randomLongitude(),
    altitude: randomAltitude(),
    frequency: randomFrequency(),
    speed: randomSpeed(),
    bearing: randomBearing(),
    ip_address: randomIpAddress(),
  };
}

/**
 * Create a random TargetUpdateDTO
 * @param partial If true, only some fields are set
 */
export function createRandomTargetUpdate(partial = true): TargetUpdateDTO {
  if (!partial) {
    return {
      latitude: randomLatitude(),
      longitude: randomLongitude(),
      altitude: randomAltitude(),
      frequency: randomFrequency(),
      speed: randomSpeed(),
      bearing: randomBearing(),
      ip_address: randomIpAddress(),
    };
  }

  const update: TargetUpdateDTO = {};

  if (Math.random() < 0.5) update.latitude = randomLatitude();
  if (Math.random() < 0.5) update.longitude = randomLongitude();
  if (Math.random() < 0.5) update.altitude = randomAltitude();
  if (Math.random() < 0.5) update.frequency = randomFrequency();
  if (Math.random() < 0.5) update.speed = randomSpeed();
  if (Math.random() < 0.5) update.bearing = randomBearing();
  if (Math.random() < 0.5) update.ip_address = randomIpAddress();

  // Ensure at least one field is set
  if (Object.keys(update).length === 0) {
    update.speed = randomSpeed();
  }

  return update;
}

/**
 * Create a list of random TargetDTOs
 */
export function createRandomTargets(count = 5): TargetDTO[] {
  return Array.from({ length: count }, () => createRandomTarget());
}
